File: java_src/src/main/java/com/google/crypto/tink/jwt/Jwt.java
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.jwt;

import com.google.errorprone.annotations.Immutable;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * A read-only implementation of <a href="https://tools.ietf.org/html/rfc7519">JSON Web Token</a>
 * (JWT).
 *
 * <p>A new instance of this class is returned as the result of a sucessfully verification of a JWT.
 *
 * @see {@link JwsMac#verifyMac}
 */
@Immutable
public final class Jwt {

  @SuppressWarnings("Immutable") // We do not mutate the header.
  private final JSONObject header;

  @SuppressWarnings("Immutable") // We do not mutate the payload.
  private final JSONObject payload;

  @SuppressWarnings("Immutable") // We do not mutate the clock.
  private final Clock clock;

  private final Duration clockSkew;

  Jwt(JSONObject header, JSONObject payload, Clock clock, Duration clockSkew) {
    this.header = header;
    this.payload = payload;
    this.clock = clock;
    this.clockSkew = clockSkew;
  }

  String getHeader(String name) throws JwtExpiredException, JwtNotBeforeException {
    validateTimestampClaims();

    try {
      return header.getString(name);
    } catch (JSONException ex) {
      return null;
    }
  }

  /**
   * Returns the {@code iss} claim that identifies the principal that issued the JWT or {@code null}
   * for none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getIssuer() throws JwtExpiredException, JwtNotBeforeException {
    return (String) getClaim(JwtNames.CLAIM_ISSUER);
  }

  /**
   * Returns the {@code sub} claim identifying the principal that is the subject of the JWT or
   * {@code null} for none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getSubject() throws JwtExpiredException, JwtNotBeforeException {
    return (String) getClaim(JwtNames.CLAIM_SUBJECT);
  }

  /**
   * Returns the {@code aud} claim identifying the principals that are the audience of the JWT or
   * {@code null} for none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public List<String> getAudiences() throws JwtExpiredException, JwtNotBeforeException {
    JSONArray audiences = (JSONArray) getClaim(JwtNames.CLAIM_AUDIENCE);
    if (audiences == null) {
      return null;
    }

    List<String> result = new ArrayList<>(audiences.length());
    for (int i = 0; i < audiences.length(); i++) {
      try {
        result.add(audiences.getString(i));
      } catch (JSONException ex) {
        throw new IllegalStateException("invalid audience", ex);
      }
    }

    return Collections.unmodifiableList(result);
  }

  /**
   * Returns the {@code jti} claim that provides a unique identifier for the JWT or {@code null} for
   * none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getJwtId() throws JwtExpiredException, JwtNotBeforeException {
    return (String) getClaim(JwtNames.CLAIM_JWT_ID);
  }

  /**
   * Returns the expiration time claim {@code exp} that identifies the instant on or after which the
   * token MUST NOT be accepted for processing or {@code null} for none.
   *
   * <p>This API requires {@link java.time.Instant} which is unavailable on Android until API level
   * 26. To use it on older Android devices, enable API desugaring as shown in
   * https://developer.android.com/studio/write/java8-support#library-desugaring.
   */
  public Instant getExpiration() {
    return getInstant(JwtNames.CLAIM_EXPIRATION);
  }

  /**
   * Returns the not before claim {@code nbf} that identifies the instant before which the token
   * MUST NOT be accepted for processing or {@code null} for none.
   *
   * <p>This API requires {@link java.time.Instant} which is unavailable on Android until API level
   * 26. To use it on older Android devices, enable API desugaring as shown in
   * https://developer.android.com/studio/write/java8-support#library-desugaring.
   */
  public Instant getNotBefore() {
    return getInstant(JwtNames.CLAIM_NOT_BEFORE);
  }

  /**
   * Returns the issued at time claim {@code iat} that identifies the instant at which the JWT was
   * issued or {@code null} for none.
   *
   * <p>This API requires {@link java.time.Instant} which is unavailable on Android until API level
   * 26. To use it on older Android devices, enable API desugaring as shown in
   * https://developer.android.com/studio/write/java8-support#library-desugaring.
   */
  public Instant getIssuedAt() {
    return getInstant(JwtNames.CLAIM_ISSUED_AT);
  }

  /**
   * Returns the claim of name {@code name} or {@code null} for none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public Object getClaim(String name) throws JwtExpiredException, JwtNotBeforeException {
    validateTimestampClaims();

    try {
      return payload.get(name);
    } catch (JSONException ex) {
      return null;
    }
  }

  /**
   * Returns the type of this JWT.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getType() throws JwtExpiredException, JwtNotBeforeException {
    return getHeader(JwtNames.HEADER_TYPE);
  }

  /**
   * Returns the content type header parameter used to declare structural information about the JWT.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getContentType() throws JwtExpiredException, JwtNotBeforeException {
    return getHeader(JwtNames.HEADER_CONTENT_TYPE);
  }

  /**
   * Returns the name of the algorithm used to sign or authenticate the JWT.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getAlgorithm() throws JwtExpiredException, JwtNotBeforeException {
    validateTimestampClaims();

    try {
      return header.getString(JwtNames.HEADER_ALGORITHM);
    } catch (JSONException ex) {
      throw new IllegalStateException("an alg header is required, but not found", ex);
    }
  }

  /**
   * Returns the ID of the key used to sign or authenticate the JWT or {@code null} for none.
   *
   * @throws JwtExpiredException when this token has been expired
   * @throws JwtNotBeforeException when this token can't be used yet
   */
  public String getKeyId() throws JwtExpiredException, JwtNotBeforeException {
    return getHeader(JwtNames.HEADER_KEY_ID);
  }

  private Instant getInstant(String name) {
    try {
      return Instant.ofEpochSecond(payload.getLong(name));
    } catch (JSONException ex) {
      return null;
    }
  }

  private void validateTimestampClaims() throws JwtExpiredException, JwtNotBeforeException {
    Instant now = this.clock.instant();

    Instant exp = this.getExpiration();
    if (exp != null && exp.isBefore(now.minus(this.clockSkew))) {
      throw new JwtExpiredException("token has expired since " + exp);
    }

    Instant nbf = this.getNotBefore();
    if (nbf != null && nbf.isAfter(now.plus(this.clockSkew))) {
      throw new JwtNotBeforeException("token cannot be used before " + nbf);
    }
  }
}

File: cc/jwt/jwt.cc
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/jwt/jwt.h"

#include "absl/strings/numbers.h"
#include "absl/strings/substitute.h"
#include "tink/jwt/jwt_names.h"

namespace crypto {
namespace tink {
util::StatusOr<std::unique_ptr<Jwt>> Jwt::New(
    const google::protobuf::Struct& header,
    const google::protobuf::Struct& payload, const absl::Time clock,
    const absl::Duration clockSkew) {
  std::unique_ptr<Jwt> jwt(new Jwt(header, payload, clock, clockSkew));
  return jwt;
}

Jwt::Jwt(const google::protobuf::Struct& header,
         const google::protobuf::Struct& payload, const absl::Time clock,
         const absl::Duration clockSkew) {
  this->header_ = header;
  this->payload_ = payload;
  this->clock_skew_ = clockSkew;
  this->clock_ = clock;
}

util::StatusOr<std::string> Jwt::GetValueAsString(
    const google::protobuf::Struct& j_proto, absl::string_view name) const {
  const auto& it = j_proto.fields().find(std::string(name));
  if (it == j_proto.fields().cend()) {
    return crypto::tink::util::Status(
        util::error::NOT_FOUND, absl::Substitute("field '$0' not found", name));
  }

  if (it->second.kind_case() != google::protobuf::Value::kStringValue) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field $0' is not a string", name));
  }

  return it->second.string_value();
}

util::StatusOr<double> Jwt::GetValueAsNumber(
    const google::protobuf::Struct& j_proto, absl::string_view name) const {
  const auto& it = j_proto.fields().find(std::string(name));
  if (it == j_proto.fields().cend()) {
    return crypto::tink::util::Status(
        util::error::NOT_FOUND, absl::Substitute("field '$0' not found", name));
  }

  if (it->second.kind_case() != google::protobuf::Value::kNumberValue) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is not a number", name));
  }

  return it->second.number_value();
}

util::StatusOr<absl::Time> Jwt::GetValueAsTime(
    const google::protobuf::Struct& j_proto, absl::string_view name) const {
  auto number_or = GetValueAsNumber(j_proto, name);
  if (!number_or.status().ok()) {
    return number_or.status();
  }

  return absl::FromUnixSeconds(number_or.ValueOrDie());
}

util::StatusOr<const google::protobuf::ListValue*> Jwt::GetValueAsList(
    const google::protobuf::Struct& j_proto, absl::string_view name) const {
  const auto& it = j_proto.fields().find(std::string(name));
  if (it == j_proto.fields().cend()) {
    return crypto::tink::util::Status(
        util::error::NOT_FOUND, absl::Substitute("field '$0' not found", name));
  }

  if (it->second.kind_case() != google::protobuf::Value::kListValue) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is not a list", name));
  }

  return static_cast<const google::protobuf::ListValue*>(
      &it->second.list_value());
}

util::StatusOr<std::vector<std::string>> Jwt::GetClaimAsStringList(
    absl::string_view name) const {
  std::vector<std::string> vec;
  auto list_or = GetValueAsList(this->payload_, name);
  if (!list_or.status().ok()) {
    return list_or.status();
  }

  auto& list = list_or.ValueOrDie();
  for (const auto& v : list->values()) {
    if (v.kind_case() != google::protobuf::Value::kStringValue) {
      return crypto::tink::util::Status(
          util::error::INVALID_ARGUMENT,
          absl::Substitute(
              "field '$0' contains an element that is not a string", name));
    }
    vec.push_back(v.string_value());
  }

  if (vec.empty()) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is empty", name));
  }

  return vec;
}

util::StatusOr<std::vector<double>> Jwt::GetClaimAsNumberList(
    absl::string_view name) const {
  std::vector<double> vec;
  auto list_or = GetValueAsList(this->payload_, name);
  if (!list_or.status().ok()) {
    return list_or.status();
  }

  auto list = list_or.ValueOrDie();
  for (const auto& v : list->values()) {
    if (v.kind_case() != google::protobuf::Value::kNumberValue) {
      return crypto::tink::util::Status(
          util::error::INVALID_ARGUMENT,
          absl::Substitute(
              "field '$0' contains an element that is not a number", name));
    }
    vec.push_back(v.number_value());
  }

  if (vec.empty()) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is empty", name));
  }

  return vec;
}

util::StatusOr<std::vector<std::string>> Jwt::GetAudiences() const {
  std::vector<std::string> vec;

  auto aud_or = GetValueAsString(this->payload_, JwtNames::kClaimAudience);
  if (aud_or.status().ok()) {
    vec.push_back(aud_or.ValueOrDie());
    return vec;
  }

  auto aud_list_or = GetValueAsList(this->payload_, JwtNames::kClaimAudience);
  if (!aud_list_or.status().ok()) {
    return aud_list_or.status();
  }

  auto aud_list = aud_list_or.ValueOrDie();
  for (const auto& v : aud_list->values()) {
    if (v.kind_case() != google::protobuf::Value::kStringValue) {
      return crypto::tink::util::Status(
          util::error::INVALID_ARGUMENT,
          absl::Substitute(
              "field '$0' contains an element that is not a string",
              JwtNames::kClaimAudience));
    }
    vec.push_back(v.string_value());
  }

  if (vec.empty()) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is empty", JwtNames::kClaimAudience));
  }

  return vec;
}

util::StatusOr<bool> Jwt::GetValueAsBool(
    const google::protobuf::Struct& j_proto, absl::string_view name) const {
  const auto& it = j_proto.fields().find(std::string(name));
  if (it == j_proto.fields().cend()) {
    return crypto::tink::util::Status(
        util::error::NOT_FOUND, absl::Substitute("field '$0' not found", name));
  }

  if (it->second.kind_case() != google::protobuf::Value::kBoolValue) {
    return crypto::tink::util::Status(
        util::error::INVALID_ARGUMENT,
        absl::Substitute("field '$0' is not a list", name));
  }

  return it->second.bool_value();
}

util::StatusOr<double> Jwt::GetClaimAsNumber(absl::string_view name) const {
  return GetValueAsNumber(this->payload_, name);
}

util::StatusOr<bool> Jwt::GetClaimAsBool(absl::string_view name) const {
  return GetValueAsBool(this->payload_, name);
}

util::StatusOr<std::string> Jwt::GetSubject() const {
  return GetValueAsString(this->payload_, JwtNames::kClaimSubject);
}

util::StatusOr<absl::Time> Jwt::GetExpiration() const {
  return GetValueAsTime(this->payload_, JwtNames::kClaimExpiration);
}

util::StatusOr<absl::Time> Jwt::GetNotBefore() const {
  return GetValueAsTime(this->payload_, JwtNames::kClaimNotBefore);
}

util::StatusOr<absl::Time> Jwt::GetIssuedAt() const {
  return GetValueAsTime(this->payload_, JwtNames::kClaimIssuedAt);
}

util::StatusOr<std::string> Jwt::GetIssuer() const {
  return GetValueAsString(this->payload_, JwtNames::kClaimIssuer);
}

util::StatusOr<std::string> Jwt::GetJwtId() const {
  return GetValueAsString(this->payload_, JwtNames::kClaimJwtId);
}

util::StatusOr<std::string> Jwt::GetContentType() const {
  return GetValueAsString(this->header_, JwtNames::kHeaderContentType);
}

util::StatusOr<std::string> Jwt::GetAlgorithm() const {
  return GetValueAsString(this->header_, JwtNames::kHeaderAlgorithm);
}

util::StatusOr<std::string> Jwt::GetKeyId() const {
  return GetValueAsString(this->header_, JwtNames::kHeaderKeyId);
}

util::StatusOr<std::string> Jwt::GetClaimAsString(
    absl::string_view name) const {
  return GetValueAsString(this->payload_, name);
}

util::StatusOr<std::string> Jwt::GetType() const {
  return GetValueAsString(this->header_, JwtNames::kHeaderType);
}

util::Status Jwt::validateTimestampClaims() const {
  absl::Time now = this->clock_;

  auto exp_or = GetExpiration();
  if (!exp_or.status().ok() &&
      exp_or.status().error_code() != util::error::NOT_FOUND) {
    // This is an error, e.g. malformated input
    return exp_or.status();
  }

  if (exp_or.status().ok()) {
    // Validate the expiration.
    auto exp = exp_or.ValueOrDie();
    if (now > exp + this->clock_skew_) {
      return crypto::tink::util::Status(util::error::OUT_OF_RANGE,
                                        "token is expired");
    }
  }

  auto nbf_or = GetNotBefore();
  if (!nbf_or.status().ok() &&
      nbf_or.status().error_code() != util::error::NOT_FOUND) {
    // This is an error, e.g. malformated input
    return nbf_or.status();
  }

  if (nbf_or.status().ok()) {
    // Validate the nbf_or.
    auto nbf = nbf_or.ValueOrDie();
    if (now < nbf - this->clock_skew_) {
      return crypto::tink::util::Status(util::error::OUT_OF_RANGE,
                                        "token is not yet valid");
    }
  }

  return util::OkStatus();
}

}  // namespace tink
}  // namespace crypto
